### DISCUSSION TITLE
Allow Type for generic instantiation

### DISCUSSION THREAD

**Discussion by msedi** [dotnet/csharplang#5284] [Language Ideas]
I'm often having the problem that I know the `Type` but having a generic method where I want the `Type `to be returned from the generic method. Following example:

```cs
public static class Convert
{
    public static TTo[]<TFrom, TTo>ChangeType(TFrom[] data);
}

public static void Main(string[] args)
{
  float[] data = new[100];

   // Type is defined
  int[] converted = Convert.ChangeType<float, int>(data);

  // Type is coming from somewhere but is not defined as <T>
  Type toType;
  
  // Proposal: allow to use the Type directly.
  int[] converted = Convert.ChangeType<float, toType>(data);
}
```

There are currently a few options to solve this, `MakeGenericMethod `is maybe the nearest solution. Another solution would be to offer an untyped `ChangeType `method, as done in `System.Convert`, but then you have to if..then all types you might ever know. Especially when C# introduces new types (like the `Half`), it can cause a lot of pain to dig through the necessary parts in a larger codebase, see here:

```cs
public static class Convert
{
    public static Array ChangeType(Array data, Type targetType);
}

```

While this is maybe currently the preferred way (I don't know?) it seems a little outdated to me since generics have been introduced. So my question is, if something like proposed above would be possible? I don't really know how the runtime treats `<T>` vs. `Type`. If it treats them identical the runtime could also create variants of the Type as shown above.

**Comment by huoyaoyuan** [comment]
Note: `ChangeType` is a quite bad sample. The conversion has to be aware of the details of two types, and then specialize for each combination.

` I don't really know how the runtime treats `<T>` vs. `Type`. 

They are fundamentally different. `<T>` is a constant to JIT and can be used for type checking etc. `Type` is a variable and you can't create a variable with the type of `Type` object. The generated code must either:
- Generate different code for each `Type` get passed in. This is similar to call `MakeGenericMethod`.
- Perform runtime type check afterwards. This is similar to use `object` and do a cast.

**Reply by msedi** [reply]
@huoyaoyuan: Thanks. That what I currently already do, but to create different code, the efforts are extremely high and the problem is maintainability. Currently we have very fast convert routines from one type into another (mostly primitive types), because the runtime does not support this very well. There is of course Array.ConvertAll, but the performance is not what we need.

So we created routines from ~12 types into another ~12 types. This means that I have to write 144 methods with all the same code,

Doing runtime checks is exactly the problem because here we a the performance bottleneck. I cannot give you detailed performance numbers because the library exists for a very long time. But if needed I can check if I can get some numbers.

**Comment by fa1ntl1ght** [comment]
#5267? 🤔

**Reply by msedi** [reply]
@0x0737. Thanks for pointing to this. I think it's similar.
